### PR TITLE
Creating a logger pattern

### PR DESCRIPTION
**Pull request by andreev-io**

### CODE REVIEW

File: Airtap/Providers/LogProvider.swift [hunk 1]
```diff
@@ -0,0 +1,38 @@
...
+//  logProvider.swift
+//  Airtap
+//
+//  Created by Ilya Andreev on 7/17/21.
+//  Copyright © 2021 Airtap Ltd. All rights reserved.
+//
+
+import Foundation
+
+enum logLevel: Equatable {
```
Comment (alxdrlitreev) [inline Airtap/Providers/LogProvider.swift]: `enums` names should start with uppercase letter in Swift

File: Airtap/Providers/LogProvider.swift [hunk 2]
```diff
@@ -0,0 +1,38 @@
...
+}
+
+protocol LogProviding {
+    var log: [LogEntry] { get }
+    
+    func addLogEntry(entry: LogEntry)
+}
+
+class LogProvider: LogProviding {
+    private(set) var log: [LogEntry] = []
```
Comment (alxdrlitreev) [inline Airtap/Providers/LogProvider.swift]: There is no need for `log` to be visible outside of the class. Please make it strictly private and remove it from protocol.

**Review by alxdrlitreev — CHANGES_REQUESTED** [review]
Good, but need some minor adjustments

File: Airtap/Providers/LogProvider.swift (hunk 1 above)
Comment (andreev-io) [reply Airtap/Providers/LogProvider.swift]: Fixed

File: Airtap/Providers/LogProvider.swift (hunk 2 above)
Comment (andreev-io) [reply Airtap/Providers/LogProvider.swift]: Fixed

**Comment by andreev-io** [comment]
I didn't add logs _everywhere_, only to some places. I suggest we add later them as we find necessary.

**Comment by andreev-io** [comment]
Updated the PR, leaving it to you to integrate with Sentry @alxdrlitreev

File: Airtap/Network/API/APIService.swift [hunk 3]
```diff
@@ -60,6 +62,7 @@ class APIService: APIServing {
     }
     
     func createAccount(licenseKey: String, firstName: String, lastName: String?) -> Future<CreateAccountResponse, NetworkingError> {
+        self.logProvider.add(LogLevel.debug, String(format: "creating a new account"))
```
Comment (alxdrlitreev) [inline Airtap/Network/API/APIService.swift]: Don't get why we need formatter here. Simply replace with:
`self.logProvider.add(.debug, "creating a new account")`

Comment (alxdrlitreev) [reply Airtap/Network/API/APIService.swift]: `LogLevel` is also redundant. Just `.debug`.

File: Airtap/Network/WS/WSService.swift [hunk 4]
```diff
@@ -78,6 +80,7 @@ class WSService: WSServing {
     }
 
     func sendOffer(to accountId: Int, sdp: String) {
+        self.logProvider.add(LogLevel.debug, String(format: "account %d sending an offer to account %d", self.authProvider.accountId!, accountId))
```
Comment (alxdrlitreev) [inline Airtap/Network/WS/WSService.swift]: Replace `self.authProvider.accountId!` with `self.authProvider.accountId ?? "UNKNOWN"`. And `LogLevel.debug` with just `.debug`.

File: Airtap/Network/WS/WSService.swift [hunk 5]
```diff
@@ -88,6 +91,7 @@ class WSService: WSServing {
     }
     
     func sendAnswer(to accountId: Int, sdp: String) {
+        self.logProvider.add(LogLevel.debug, String(format: "account %d sending an answer to account %d", self.authProvider.accountId!, accountId))
```
Comment (alxdrlitreev) [inline Airtap/Network/WS/WSService.swift]: Replace `self.authProvider.accountId!` with `self.authProvider.accountId ?? "UNKNOWN"`. And `LogLevel.debug` with just `.debug`.

File: Airtap/Network/WS/WSService.swift [hunk 6]
```diff
@@ -98,6 +102,7 @@ class WSService: WSServing {
     }
     
     func sendCandidate(to accountId: Int, sdp: String, sdpMLineIndex: Int32, sdpMid: String?) {
+        self.logProvider.add(LogLevel.debug, String(format: "account %d sending a candidate to account %d", self.authProvider.accountId!, accountId))
```
Comment (alxdrlitreev) [inline Airtap/Network/WS/WSService.swift]: Replace `self.authProvider.accountId!` with `self.authProvider.accountId ?? "UNKNOWN"`. And `LogLevel.debug` with just `.debug`.

File: Airtap/Network/WS/WSService.swift [hunk 7]
```diff
@@ -108,6 +113,7 @@ class WSService: WSServing {
     }
     
     func sendInfo(to accountId: Int, type: WSPayloadInfoType) {
+        self.logProvider.add(LogLevel.debug, String(format: "account %d sending info to account %d", self.authProvider.accountId!, accountId))
```
Comment (alxdrlitreev) [inline Airtap/Network/WS/WSService.swift]: Replace `self.authProvider.accountId!` with `self.authProvider.accountId ?? "UNKNOWN"`. And `LogLevel.debug` with just `.debug`.

File: Airtap/Network/WS/WSService.swift [hunk 8]
```diff
@@ -118,6 +124,7 @@ class WSService: WSServing {
     }
     
     private func sendAck(nonce: Int) {
+        self.logProvider.add(LogLevel.debug, String(format: "account %d sending %d ack", self.authProvider.accountId!, nonce))
```
Comment (alxdrlitreev) [inline Airtap/Network/WS/WSService.swift]: Replace `self.authProvider.accountId!` with `self.authProvider.accountId ?? "UNKNOWN"`. And `LogLevel.debug` with just `.debug`.

File: Airtap/Network/WS/WSService.swift [hunk 9]
```diff
@@ -149,8 +156,10 @@ class WSService: WSServing {
         
         switch(message.type) {
         case .ack:
+            self.logProvider.add(LogLevel.debug, String(format: "account %d received an ack", self.authProvider.accountId!))
```
Comment (alxdrlitreev) [inline Airtap/Network/WS/WSService.swift]: Replace `self.authProvider.accountId!` with `self.authProvider.accountId ?? "UNKNOWN"`. And `LogLevel.debug` with just `.debug`.

File: Airtap/Network/WS/WSService.swift [hunk 10]
```diff
@@ -149,8 +156,10 @@ class WSService: WSServing {
...
             break
         case .offer:
+            self.logProvider.add(LogLevel.debug, String(format: "account %d received an offer from %d", self.authProvider.accountId!, accountId))
```
Comment (alxdrlitreev) [inline Airtap/Network/WS/WSService.swift]: Replace `self.authProvider.accountId!` with `self.authProvider.accountId ?? "UNKNOWN"`. And `LogLevel.debug` with just `.debug`.

File: Airtap/Network/WS/WSService.swift [hunk 11]
```diff
@@ -161,6 +170,7 @@ class WSService: WSServing {
                 )
             }
         case .answer:
+            self.logProvider.add(LogLevel.debug, String(format: "account %d received an answer from %d", self.authProvider.accountId!, accountId))
```
Comment (alxdrlitreev) [inline Airtap/Network/WS/WSService.swift]: Replace `self.authProvider.accountId!` with `self.authProvider.accountId ?? "UNKNOWN"`. And `LogLevel.debug` with just `.debug`.

File: Airtap/Network/WS/WSService.swift [hunk 12]
```diff
@@ -171,6 +181,7 @@ class WSService: WSServing {
                 )
             }
         case .candidate:
+            self.logProvider.add(LogLevel.debug, String(format: "account %d received a candidate from %d", self.authProvider.accountId!, accountId))
```
Comment (alxdrlitreev) [inline Airtap/Network/WS/WSService.swift]: Replace `self.authProvider.accountId!` with `self.authProvider.accountId ?? "UNKNOWN"`. And `LogLevel.debug` with just `.debug`.

File: Airtap/Network/WS/WSService.swift [hunk 13]
```diff
@@ -183,6 +194,7 @@ class WSService: WSServing {
                 )
             }
         case .info:
+            self.logProvider.add(LogLevel.debug, String(format: "account %d received info from %d", self.authProvider.accountId!, accountId))
```
Comment (alxdrlitreev) [inline Airtap/Network/WS/WSService.swift]: Replace `self.authProvider.accountId!` with `self.authProvider.accountId ?? "UNKNOWN"`. And `LogLevel.debug` with just `.debug`.

File: Airtap/Network/WS/WSService.swift [hunk 14]
```diff
@@ -192,6 +204,8 @@ class WSService: WSServing {
                     )
                 )
             }
+        case .peers:
+            self.logProvider.add(LogLevel.debug, String(format: "account %d received peers \(message.payload as Optional)", self.authProvider.accountId!))
```
Comment (alxdrlitreev) [inline Airtap/Network/WS/WSService.swift]: Replace `self.authProvider.accountId!` with `self.authProvider.accountId ?? "UNKNOWN"`. And `LogLevel.debug` with just `.debug`.

Also, `message.payload as Optional` is unacceptable.

File: Airtap/Network/WS/WSService.swift [hunk 15]
```diff
@@ -204,8 +218,8 @@ extension WSService: WebSocketDelegate {
         case .connected:
             self.eventSubject.send(.ready)
         case let .text(receivedString):
-            print("SOCKETS IN: \(receivedString)")
             let payload = try! JSONDecoder().decode(WSPayload.self, from: receivedString.data(using: .utf8)!)
+            logProvider.add(LogLevel.debug, "\(payload)")
```
Comment (alxdrlitreev) [inline Airtap/Network/WS/WSService.swift]: Replace `LogLevel.debug` with just  `.debug`.

File: Airtap/Providers/AuthProvider.swift [hunk 16]
```diff
@@ -49,6 +51,7 @@ class AuthProvider: AuthProviding, ObservableObject {
     }
     
     func signIn(accountId: Int, token: String) {
+        self.logProvider.add(LogLevel.debug, String(format: "account %d signing in", accountId))
```
Comment (alxdrlitreev) [inline Airtap/Providers/AuthProvider.swift]: Same here.

File: Airtap/Providers/AuthProvider.swift [hunk 17]
```diff
@@ -58,6 +61,7 @@ class AuthProvider: AuthProviding, ObservableObject {
     }
     
     func signOut() {
+        self.logProvider.add(LogLevel.debug, String(format: "account %d signing out", self.accountId!))
```
Comment (alxdrlitreev) [inline Airtap/Providers/AuthProvider.swift]: No unwraps, just debug

File: Airtap/Providers/PersistenceProvider.swift [hunk 18]
```diff
@@ -77,6 +79,7 @@ class PersistenceProvider: PersistenceProviding, ObservableObject {
     }
     
     func insertPeer(id: Int, firstName: String, lastName: String?) {
+        self.logProvider.add(LogLevel.debug, String(format: "account %d inserted peer %d into persistence", self.authProvider.accountId!, id))
```
Comment (alxdrlitreev) [inline Airtap/Providers/PersistenceProvider.swift]: Same and everywhere else.

**Review by alxdrlitreev — CHANGES_REQUESTED** [review]

**Comment by andreev-io** [comment]
Fixed & addressed all

**Comment by alxdrlitreev** [comment]
Conflicts @andreev-io :(

**Comment by andreev-io** [comment]
Fixed. Can merge now.

**Pull request merged**